pkg/dns/dns_test.go

```diff
@@ -0,0 +1,377 @@
+	return subset
+}
+
+func newSubset() kapi.EndpointSubset {
+	subset := kapi.EndpointSubset{
+		Addresses: []kapi.EndpointAddress{},
+		Ports:     []kapi.EndpointPort{},
+	}
+	return subset
+}
+
+func assertSRVForHeadlessService(t *testing.T, kd *KubeDNS, s *kapi.Service, e *kapi.Endpoints) {
+	for _, subset := range e.Subsets {
+		for _, port := range subset.Ports {
+			records, err := kd.Records(getSRVFQDN(kd, s, port.Name), false)
+			require.NoError(t, err)
+			assertRecordPortsMatchPort(t, port.Port, records)
+			assertCNameRecordsMatchEndpointIPs(t, kd, subset.Addresses, records)
+		}
+	}
+}
+
+func assertDNSForHeadlessService(t *testing.T, kd *KubeDNS, e *kapi.Endpoints) {
+	records, err := kd.Records(getEndpointsFQDN(kd, e), false)
+	require.NoError(t, err)
+	endpoints := map[string]bool{}
+	for _, subset := range e.Subsets {
+		for _, endpointAddress := range subset.Addresses {
+			endpoints[endpointAddress.IP] = true
+		}
+	}
+	assert.Equal(t, len(endpoints), len(records))
+	for _, record := range records {
+		_, found := endpoints[record.Host]
+		assert.True(t, found)
+	}
+}
+
+func assertRecordPortsMatchPort(t *testing.T, port int32, records []skymsg.Service) {
+	for _, record := range records {
+		assert.Equal(t, port, int32(record.Port))
+	}
+}
+
+func assertCNameRecordsMatchEndpointIPs(t *testing.T, kd *KubeDNS, e []kapi.EndpointAddress, records []skymsg.Service) {
+	endpoints := map[string]bool{}
+	for _, endpointAddress := range e {
+		endpoints[endpointAddress.IP] = true
+	}
+	assert.Equal(t, len(e), len(records), "unexpected record count")
+	for _, record := range records {
+		_, found := endpoints[getIPForCName(t, kd, record.Host)]
+		assert.True(t, found, "Did not endpoint with address:%s", record.Host)
+	}
+}
+
+func getIPForCName(t *testing.T, kd *KubeDNS, cname string) string {
+	records, err := kd.Records(cname, false)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(records), "Could not get IP for CNAME record for %s", cname)
+	assert.NotNil(t, net.ParseIP(records[0].Host), "Invalid IP address %q", records[0].Host)
+	return records[0].Host
+}
+
+func assertNoDNSForHeadlessService(t *testing.T, kd *KubeDNS, s *kapi.Service) {
+	records, err := kd.Records(getServiceFQDN(kd, s), false)
+	require.Error(t, err)
+	assert.Equal(t, 0, len(records))
+}
+
+func assertSRVForNamedPort(t *testing.T, kd *KubeDNS, s *kapi.Service, portName string) {
+	records, err := kd.Records(getSRVFQDN(kd, s, portName), false)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(records))
+	assert.Equal(t, getServiceFQDN(kd, s), records[0].Host)
+}
+
+func assertNoSRVForNamedPort(t *testing.T, kd *KubeDNS, s *kapi.Service, portName string) {
+	records, err := kd.Records(getSRVFQDN(kd, s, portName), false)
+	require.Error(t, err)
+	assert.Equal(t, 0, len(records))
+}
+
+func assertNoDNSForClusterIP(t *testing.T, kd *KubeDNS, s *kapi.Service) {
+	records, err := kd.Records(getServiceFQDN(kd, s), false)
```

Comment:
Shouldn't this be verified for all the names/queries in the same way you do in assertDNSForClusterIP?